addons/mail/models/mail_render_mixin.py

```diff
@@ -413,31 +384,30 @@ def _render_template_jinja(self, template_txt, model, res_ids,
         if not template_txt:
             return results
 
-        # try to load the template
-        try:
-            jinja_env = jinja_safe_template_env if no_autoescape else jinja_template_env
-            template = jinja_env.from_string(tools.ustr(template_txt))
-        except Exception:
-            _logger.info("Failed to load template %r", template_txt, exc_info=True)
-            return results
-
         # prepare template variables
-        variables = self._render_jinja_eval_context()
+        variables = self._render_qweb_eval_context()
         if add_context:
             variables.update(**add_context)
-        safe_eval.check_values(variables)
-
 
         for record in self.env[model].browse(res_ids):
             variables['object'] = record
-            try:
-                render_result = template.render(variables)
-            except Exception as e:
-                _logger.info("Failed to render template : %s", e, exc_info=True)
-                raise UserError(_("Failed to render template : %s", e))
-            if render_result == u"False":
-                render_result = u""
-            results[record.id] = Markup(render_result)
+
+            results[record.id] = ''
+
+            instructions = parse_small_qweb(tools.ustr(template_txt))
+            instructions
+            for string, expression in instructions:
+                results[record.id] += string
```

Comment:
Nope. That way quadratic runtime lies. Juse a list and `join` it at the end.

Reply:
Changed.